README.md
# iddaa
idapython scripts, including feature as below:

1. Integrete IDA pro and gdb through idapython.
2. Identify the suspicious functions in binary by static analysis.
3. Improve to analyze CGC format.

## Install
- On Windows:
    1. Change `port` in `iddaa/idapython/rpcserver.py` to yours.
    2. Copy the file in **idapython** to the directory of IDA plugin.
      - You can execute **install.bat** if you installed IDA pro at `C:\Program Files (x86)\IDA 6.8\`.

- On Linux:
    1. `git clone https://github.com/0xddaa/iddaa.git ~/tools/iddaa`
    2. `echo "source ~/tools/iddaa/gdbscript/gdbinit" >> ~/.gdbinit`
    3. Change `HOST` and `PORT` in `iddaa/gdbscript/rpc.py` to yours.

## Features
### RPC Server
- Use symbols named in IDA pro
- View the pseudo code that decompiled by IDA Pro
- Set the breakpoints that be marked in IDA Pro
- Remote idapython script execute

### CGC Helper
- Revise syscall
    - Revise correct comment in IDA pro becuase CGC use different syscall number. 
- Automatic function naming
    - All CGC binary is static linked, stripped, and never use glibc.  
    Rename the function if identified the possible pattern.

### Why not XMLPRC?
There are obvious delay if using xmlrpc to execute idapython.  
Some functions, such as `idaapi.decompile`, will let IDA pro be stucked with no reason.  
If you know the solution, please tell me. Orz  

## Usage
### RPC Server
You can use this command in **gdb**:
- `get_ida_symbols`  
    Get all symbols named in IDA pro.
- `get_pseudo_code [function]`  
    Get the pseudo code of specified function.
- `get_local_type`  
    Get all local types defined in IDA pro.
- `get_breakpoints`  
    Get breakpoints that be marked in IDA Pro and set breakpoints automatically.
- `idapython <files>`  
    Execute file in ida pro and get result.  
- `idc|idaapi|idautils <code>`  
    The RPC wrapper of idapython. Show the cheatsheet with `idapython cheatsheet`.
- `idapython cheatsheet`  
    Show some common used fucntion in idapython.

```
idc MakeComm(addr, comment)
----------------------------------------
Add comment at specified address.
Ex: idc MakeComm(0x804ddaa, 'Soy Sauce')

idc SetColor(addr, what, color)
----------------------------------------
Set color for specified area
Ex: idc SetColor(0x0804ddaa, 1, 0xaabbcc) // address only
    idc SetColor(0x0804ddaa, 2, 0xaabbcc) // entire function
    idc SetColor(0x0804ddaa, 3, 0xaabbcc) // entire segment
```

### CGC Helper
**deprecated**
IDACGC is better.  

- Revise syscall
    - Press `Shift + R` to revise the syscall comment
    - Execute `CGCHelper.revise_syscall(True)` in console if you want to change the function name together


CHANGELOG.md
# iddaa Change Log
## [0.0.3] - 2017-07-09
### Changed
- Fix RPCServer bugs
- Deprecate CGCHelper

## [0.0.2] - 2016-06-09
### Added
- Collect breadpoints in IDA Pro

### Changed
- Universe collector name to InfoCellctor

## [0.0.1] - 2016-06-08
### Added
- Release the first version


gdbscript/rpc.py
from __future__ import print_function
import gdb
import socket
import pickle
import os
import subprocess as sp
import sys

IDA_HOST = '10.113.208.101'
PORT = 56746
TMPDIR = '/tmp/iddaa'

def connect_ida():
    if not os.path.exists(TMPDIR):
        os.mkdir(TMPDIR)
    try:
        sock = socket.create_connection((IDA_HOST, PORT), timeout=3)
        return sock
    except socket.error as err:
        sys.stderr.write("[ERROR] {}\n".format(err))
        return None

def show_result(result):
    try:
        f = open('{}/result'.format(TMPDIR), 'w')
        f.write(result)
        f.close()
    except err:
        sys.stderr.write("[ERROR] {}\n".format(''))
        return
    gdb.execute('shell vim {}/result'.format(TMPDIR))

def send(sock, buf):
    if sys.version_info < (3, 0):
        sock.send(buf)
    else:
        sock.send(bytes(buf, 'UTF-8'))

def recv(sock, raw=False):
    buf = bytes()
    while True:
        tmp = sock.recv(4096)
        buf += tmp
        if not tmp:
            break
    if raw:
        return buf
    else:
        return buf if sys.version_info < (3, 0) else buf.decode()

def get_ida_symbols():
    sock = connect_ida()
    if not sock: return

    send(sock, 'GETSYM')
    buf = recv(sock, True)

    with open('{}/symfile'.format(TMPDIR), 'wb') as f:
        f.write(buf)

    if os.path.exists('{}/symfile'.format(TMPDIR)):
        gdb.execute('symbol-file {}/symfile'.format(TMPDIR))
    else:
        print('Can\'t not receive ida symfile.')

def get_pseudo_code(func):
    sock = connect_ida()
    if not sock: return

    send(sock, 'GETPSEUDOCODE {}'.format(func))
    code = recv(sock).strip()
    if 'Function not found' in code:
        print('[Error] ' + code)
        return
    show_result(code)

def get_local_type():
    sock = connect_ida()
    if not sock: return

    send(sock, 'GETLOCALTYPE')
    buf = recv(sock, True)
    local_type = pickle.loads(buf)
    with open('{}/localtype.h'.format(TMPDIR), 'wb') as f:
        f.write(bytes(local_type['header'], 'UTF-8'))
    with open('{}/localtype.cpp'.format(TMPDIR), 'wb') as f:
        f.write(bytes(local_type['source'], 'UTF-8'))
    cwd = os.getcwd()
    os.chdir(TMPDIR)
    if sp.check_call('g++ -c -g localtype.cpp'.split(' ')) == 0:
        gdb.execute('add-symbol-file {}/localtype.o 0'.format(TMPDIR))
    else:
        print('Generate symbol file failed')
    os.chdir(cwd)

def get_breakpoints():
    sock = connect_ida()
    if not sock: return

    send(sock, 'GETBREAKPOINTS')
    buf = recv(sock, True)
    bps = pickle.loads(buf)
    print(bps)
    for bp in bps:
        gdb.execute('break *{}'.format(bp))

class IDAPYTHON(gdb.Command):
    """ IDA python script wrapper"""
    def __init__(self):
        super(IDAPYTHON, self).__init__('idapython', gdb.COMMAND_USER)
    def invoke(self, args, from_tty):
        if args == 'cheatsheet':
            self.__cheatsheet()
            return

        sock = connect_ida()
        if not sock: return

        send(sock, 'EXECFILE')
        buf = ''
        try:
            f = open(args, 'r')
            buf = f.read()
        except:
            print('[ERROR] File not found.')
            return
        send(sock, buf)
        show_result(recv(sock))

    def __cheatsheet(self):
        print('IDA python Cheat Sheet')
        print()
        print('idc MakeComm(addr, comment)')
        print('----------------------------------------')
        print('Add comment at specified address.')
        print('Ex: idc MakeComm(0x804ddaa, \'Soy Sauce\')')
        print()
        print('idc SetColor(addr, what, color)')
        print('----------------------------------------')
        print('Set color for specified area')
        print('Ex: idc SetColor(0x0804ddaa, 1, 0xaabbcc) // address only')
        print('    idc SetColor(0x0804ddaa, 2, 0xaabbcc) // entire function')
        print('    idc SetColor(0x0804ddaa, 3, 0xaabbcc) // entire segment')
        print()


class IDARPC(gdb.Command):
    """ IDA python command wrapper"""
    def __init__(self, name):
        super(IDARPC, self).__init__(name, gdb.COMMAND_USER)
        self.name = name

    def invoke(self, args, from_tty):
        sock = connect_ida()
        if not sock: return

        send(sock, 'EXEC {}.{}'.format(self.name, args))
        show_result(recv(sock))

IDAPYTHON()
IDARPC('idautils')
IDARPC('idaapi')
IDARPC('idc')
